includes/sanitizers/class-amp-layout-sanitizer.php

```diff
@@ -0,0 +1,53 @@
+<?php
+/**
+ * Class AMP_Layout_Sanitizer
+ *
+ * @since 1.5.0
+ * @package AMP
+ */
+
+/**
+ * Class AMP_Layout_Sanitizer
+ *
+ * @since 1.5.0
+ */
+class AMP_Layout_Sanitizer extends AMP_Base_Sanitizer {
+
+	/**
+	 * Sanitize any element that has the `layout` or `data-amp-layout` attribute.
+	 */
+	public function sanitize() {
+		$xpath = new DOMXPath( $this->dom );
+		$nodes = $xpath->query( '//*[ @layout or @data-amp-layout ]' );
+
+		foreach ( $nodes as $node ) {
+			$width  = $node->getAttribute( 'width' );
+			$height = $node->getAttribute( 'height' );
+
+			// The `layout` attribute can also be defined through the `data-amp-layout` attribute.
+			if ( $node->hasAttribute( 'data-amp-layout' ) ) {
+				$layout = $node->getAttribute( 'data-amp-layout' );
+				$node->setAttribute( 'layout', $layout );
+				$node->removeAttribute( 'data-amp-layout' );
+			}
+
+			if ( ! $node->hasAttribute( 'layout' ) ) {
```

Comment:
Does this make sense here? The `$node` at this point is guaranteed to _always_ have a `layout` attribute, so this code will never run.

Reply:
I see two ways to make this happen:

- Put `AMP_Layout_Sanitizer` to be the last sanitizer to run. This ensures that `AMP_Tag_And_Attribute_Sanitizer` would remove elements with invalid layouts before sanitizing the layout itself.

- Keep the current order of the sanitizers and move the logic for validating invalid layouts to `AMP_Layout_Sanitizer`.

Which would you prefer?

Reply:
Implement the first point in https://github.com/ampproject/amp-wp/pull/3728/commits/8a2b94d3ab8b58ef972db3bd1661d3cb9bc67637.

Reply:
I think the validation of the `layout` attributes should remain in the `AMP_Tag_And_Attribute_Sanitizer`, where it removes any elements with invalid `layout`. The `AMP_Layout_Sanitizer` can then be purely focused on fixing up non-AMP HTML to infer the right `layout` based on the HTML.

So I think `AMP_Tag_And_Attribute_Sanitizer` should continue to be the last sanitizer that runs.